### DISCUSSION TITLE
Using a "missing field" sentinel, problems with serialisation

### DISCUSSION THREAD

**Discussion by LewisGaul** [pydantic/pydantic#9943] [Question]
I want to write a model that is clear about whether fields were provided or omitted, e.g. for serialisation. Using `None` isn't a great solution because it's a valid value.

I tried using a sentinel class:
```python
from pathlib import Path

import pydantic

class MissingSentinel:
    _instance = None

    def __new__(cls):
        if not cls._instance:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __repr__(self):
        return "<MISSING>"

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if isinstance(v, cls):
            return v
        raise ValueError(f"Cannot instantiate {cls.__name__} with {v}")

    @classmethod
    def __get_pydantic_json_schema__(cls):
        return {"type": "string", "enum": []}

MISSING = MissingSentinel()

class MyModel(pydantic.BaseModel):
    foo: int
    bar: str | MissingSentinel = MISSING
    baz: Path | MissingSentinel = MISSING

MyModel(foo=123).model_dump(mode="json")
```

With the above code I get
```
pydantic_core._pydantic_core.PydanticSerializationError: Unable to serialize unknown type: <class 'MissingSentinel'>
```

What I tried/considered (I've been googling and reading docs for a while now):
- It works with `model_dump(mode="python")`, but I don't know how to then use pydantic to convert types to JSON-compatible (e.g. `Path`).
- Using `None` instead of a sentinel works with `model_dump(exclude_none=True)`, but then there's no way to distinguish between a field set to `None` and a field that wasn't set.
- Using `model_dump(exclude_unset=True)` works until you have a field with a default value that you do want to include, e.g. discriminated unions.
- I assume there's a way to tell pydantic how to JSON-serialise the custom sentinel type and then exclude it by using a custom `model_dump()`, but at that point it defeats the point of using a sentinel class over just using an arbitrary string which could theoretically conflict with a real value, also not sure how to robustly implement a custom `model_dump()` to exclude fields...
- I saw some mention of using something like `Field(default_factory=lambda: ...)`, but that seems very non-obvious for anyone reading the code.

Isn't this a common thing to want to do? Am I missing the standard solution?

**Comment by uriyyo** [comment]
Hi @LewisGaul,

You can try to override `__get_pydantic_core_schema__`:
```py
from pathlib import Path
from typing import Any

import pydantic
from pydantic import GetCoreSchemaHandler
from pydantic_core import core_schema


class MissingSentinel:
    _instance = None

    def __new__(cls):
        if not cls._instance:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __repr__(self):
        return "<MISSING>"

    @classmethod
    def __get_pydantic_core_schema__(cls, source_type: Any, _: GetCoreSchemaHandler) -> core_schema.CoreSchema:
        return core_schema.is_instance_schema(
            cls=source_type,
            serialization=core_schema.to_string_ser_schema(),
        )

    @classmethod
    def __get_pydantic_json_schema__(cls):
        return {"type": "string", "enum": []}


MISSING = MissingSentinel()


class MyModel(pydantic.BaseModel):
    foo: int
    bar: str | MissingSentinel = MISSING
    baz: Path | MissingSentinel = MISSING


res = MyModel(foo=123).model_dump(mode="json")
print(res)
```

**Reply by ricosaurus** [reply]
I need a missing sentinel as well, so appreciate this.  But just starting to try to understand the  current __get_pydantic_json_schema__, and 

`MyModel.model_json_schema()`

gives ...

TypeError: MissingSentinel.__get_pydantic_json_schema__() takes 1 positional argument but 3 were given.

Does V2 now takes more args?  Looking at [https://docs.pydantic.dev/latest/concepts/json_schema/#implementing-__get_pydantic_json_schema__](url).

**Comment by uriyyo** [comment]
Or you can add serializer using `Annotated`:
```py
from pathlib import Path
from typing import Annotated

import pydantic


class MissingSentinel:
    _instance = None

    def __new__(cls):
        if not cls._instance:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __repr__(self):
        return "<MISSING>"

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if isinstance(v, cls):
            return v
        raise ValueError(f"Cannot instantiate {cls.__name__} with {v}")

    @classmethod
    def __get_pydantic_json_schema__(cls):
        return {"type": "string", "enum": []}


MISSING = MissingSentinel()

SerializableMissingSentinel = Annotated[
    MissingSentinel,
    pydantic.PlainSerializer(str, return_type=str),
]


class MyModel(pydantic.BaseModel):
    foo: int
    bar: str | SerializableMissingSentinel = MISSING
    baz: Path | SerializableMissingSentinel = MISSING


r = MyModel(foo=123).model_dump(mode="json")
print(r)

```

**Comment by LewisGaul** [comment]
@uriyyo thanks for the response, it seems you're offering a couple of mechanisms for making the sentinel class serialisable, which corresponds to my point:

> * I assume there's a way to tell pydantic how to JSON-serialise the custom sentinel type and then exclude it by using a custom `model_dump()`, but at that point it defeats the point of using a sentinel class over just using an arbitrary string which could theoretically conflict with a real value, also not sure how to robustly implement a custom `model_dump()` to exclude fields...

In other words, this feels slightly better than just using `None`, but not much better than just using an arbitrary string as a sentinel - how would I then distinguish between arbitrary string values and missing fields?

**Comment by Viicos** [comment]
A new `MISSING` sentinel was introduced in https://github.com/pydantic/pydantic/pull/11883 as an experimental feature.

It relies on the draft [PEP 661](https://peps.python.org/pep-0661/), introducing sentinels into the standard library (that is why this feature is marked as experimental).

`MISSING` was used as a name instead of `UNSET`, to avoid confusion with the current concept of "unset" fields (such fields are tracked _implicitly_ by Pydantic, and are only excluded when setting `exclude_unset=True`).

During serialization, any field with `MISSING` as a value is excluded from the output, and the sentinel doesn't appear in the JSON Schema, making it an implementation detail of your model, not exposed to external consumers.

```python
from pydantic import BaseModel
from pydantic.experimental.missing_sentinel import MISSING

class Configuration(BaseModel):
    timeout: int | None | MISSING = MISSING

conf = Configuration()

conf.model_dump()
# {}
Configuration.model_json_schema()['properties']['timeout']
#> {'anyOf': [{'type': 'integer'}, {'type': 'null'}], 'title': 'Timeout'}}
```

Sentinels are [special-cased by static type checkers](https://peps.python.org/pep-0661/#typing), meaning there is no need for any kind of `MissingType` in annotations. This feature is only supported by Pyright as of today.

See [the documentation](https://docs.pydantic.dev/dev/concepts/experimental/#missing-sentinel) for more details.

Stabilization of this feature can be tracked in [this issue](https://github.com/pydantic/pydantic/issues/12090).
